In [ ]:
#Установка нужных пакетов
!pip install --upgrade nltk gensim bokeh umap-learn

import itertools
import string

import numpy as np
import umap
from nltk.tokenize import WordPunctTokenizer

from matplotlib import pyplot as plt

from IPython.display import clear_output

In [ ]:
# выгружаем датасет:
!wget https://www.dropbox.com/s/obaitrix9jyu84r/quora.txt?dl=1 -O ./quora.txt -nc

File ‘./quora.txt’ already there; not retrieving.


In [ ]:
data = list(open("./quora.txt", encoding="utf-8"))
data[50]

"What TV shows or books help you read people's body language?\n"

In [ ]:
tokenizer = WordPunctTokenizer()

print(tokenizer.tokenize(data[50]))

['What', 'TV', 'shows', 'or', 'books', 'help', 'you', 'read', 'people', "'", 's', 'body', 'language', '?']


#Задание 1: Перевести все слова в нижний регистр (NLTK) из data и добавьте как лист токенов в листе data_tok


In [ ]:
data_tok = []
for line in data:
  lst = [x.lower() for x in tokenizer.tokenize(line)]
  data_tok.append(lst)


#checking

assert all(isinstance(row, (list, tuple)) for row in data_tok), "please convert each line into a list of tokens (strings)"
is_latin = lambda tok: all('a' <= x.lower() <= 'z' for x in tok)
assert all(map(lambda l: not is_latin(l) or l.islower(), map(' '.join, data_tok))), "please make sure to lowercase the data"


###Задание 2: Подсчитайте топ10 самых популярных лем в рамках data

In [ ]:
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()

dict_words = {}

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [ ]:
flatten_data_tok = [x for sublist in data_tok for x in sublist]

In [ ]:
lemm_lst = []
for world in flatten_data_tok:
  lemm_lst.append(lemmatizer.lemmatize(world))

In [ ]:
from collections import Counter
import pandas as pd

In [ ]:
count_lemm = Counter(lemm_lst)

In [ ]:
df = pd.DataFrame.from_dict(count_lemm, orient='index')

In [ ]:
df.sort_values(0).tail(10)

0
of    112001
how   135687
in    139813
to    141788
i     149868
a     172513
is    185392
what  214798
the   252068
?     552413

###Задание 3: Подсчитайте количество разных слов до и после лемматизации

In [ ]:
len(set(flatten_data_tok)) # до лематизации и стеминга

87820

In [ ]:
len(set(lemm_lst)) # после лематизации

80304

###Задание 4: Подсчитайте количество разных слов до и после стемминга

In [ ]:
from nltk.stem.porter import PorterStemmer

In [ ]:
stemmer = PorterStemmer()

In [ ]:
stemm_lst = []
for word in flatten_data_tok:
  stemm_lst.append(stemmer.stem(word))

In [ ]:
len(set(stemm_lst)) # после стеминга

67026

###Задание 5: Подсчитайте количество разных слов



In [ ]:
lemm_stemm_lst = []
for word in lemm_lst:
  lemm_stemm_lst.append(stemmer.stem(word))

In [ ]:
stemm_lemm_lst = []
for word in stemm_lst:
  stemm_lemm_lst.append(lemmatizer.lemmatize(word))

In [ ]:
len(set(lemm_stemm_lst)) # после леминга и затем стеминга

66835

In [ ]:
len(set(stemm_lemm_lst)) # после стеминга и затем леминга

66818

**Вывод:** при использовании стемминга вперед лематизации может потеряться часть уникальных слов, нужно делать сначала лематизацию а затем стемминг

REGEXP

https://www.programiz.com/python-programming/regex 

https://docs.python.org/3/howto/regex.html

In [ ]:
import re

pattern = 'a*s'
test_string = 'abyss'
result = re.match(pattern, test_string)

if result:
  print("Search successful.")
else:
  print("Search unsuccessful.")	

Search unsuccessful.


In [ ]:
import re

pattern = '[a-zA-Z0-9]{0,20}@[a-z]{0,5}.[a-z]{0,3}'
test_string = 'garafutdinov@list.ru'
result = re.match(pattern, test_string)

if result:
  print("Search successful.")
else:
  print("Search unsuccessful.")	

Search successful.


###Задание 6: 
https://www.hackerrank.com/challenges/matching-specific-string/problem?isFullScreen=true 

###Задание 7: 
https://www.hackerrank.com/challenges/matching-whitespace-non-whitespace-character/problem?isFullScreen=true

###Задание 8: 
https://www.hackerrank.com/challenges/matching-start-end/problem?isFullScreen=true

###Задание 9: 
https://www.hackerrank.com/challenges/matching-word-boundaries/problem?isFullScreen=true

Bag Of Words (BOW)

In [ ]:
def vectorize(tokens):
    ''' This function takes list of words in a sentence as input 
    and returns a vector of size of filtered_vocab.It puts 0 if the 
    word is not present in tokens and count of token if present.'''
    vector=[]
    for w in filtered_vocab:
        vector.append(tokens.count(w))
    return vector
def unique(sequence):
    '''This functions returns a list in which the order remains 
    same and no item repeats.Using the set() function does not 
    preserve the original ordering,so i didnt use that instead'''
    seen = set()
    return [x for x in sequence if not (x in seen or seen.add(x))]

#create a list of stopwords.You can import stopwords from nltk too
stopwords=["to","is","a"]

#list of special characters.You can use regular expressions too
special_char=[",",":"," ",";",".","?"]

#Write the sentences in the corpus,in our case, just two 
string1="Welcome to Great Learning , Now start learning"
string2="Learning Learning is a good practice"

#convert them to lower case
string1=string1.lower()
string2=string2.lower()

#split the sentences into tokens
tokens1=string1.split()
tokens2=string2.split()
print(tokens1)
print(tokens2)

#create a vocabulary list
vocab=unique(tokens1+tokens2)
print(vocab)

#filter the vocabulary list
filtered_vocab=[]
for w in vocab: 
    if w not in stopwords and w not in special_char: 
        filtered_vocab.append(w)
print(filtered_vocab)

#convert sentences into vectords
vector1=vectorize(tokens1)
print(vector1)
vector2=vectorize(tokens2)
print(vector2)

['welcome', 'to', 'great', 'learning', ',', 'now', 'start', 'learning']
['learning', 'learning', 'is', 'a', 'good', 'practice']
['welcome', 'to', 'great', 'learning', ',', 'now', 'start', 'is', 'a', 'good', 'practice']
['welcome', 'great', 'learning', 'now', 'start', 'good', 'practice']
[1, 1, 2, 1, 1, 0, 0]
[0, 0, 2, 0, 0, 1, 1]


Задание 10: Реализовать Bag of words на data_tok (можно на NLTK, можно без)

In [ ]:
first50 = ''
for line in data[0:50]:
  first50 += line
first50

"Can I get back with my ex even though she is pregnant with another guy's baby?\nWhat are some ways to overcome a fast food addiction?\nWho were the great Chinese soldiers and leaders who fought in WW2?\nWhat are ZIP codes in the Bay Area?\nWhy was George RR Martin critical of JK Rowling after losing the Hugo award?\nWhat can I do to improve my immune system?\nHow is your relationship with your mother in law?\nHow does one get Free PSN codes/Vita Codes?\nWhat is your review of osquery?\nHow can I look smart and act smart?\nWhich brand should go with the GTX 960 graphic card, MSI, Zotac or ASUS?\nWhat is the ZIP code of India?\nAs an Indian doctor practicing in my own private clinic. How can I earn more? What other sources of income can I use?\nIs Windows the only OS that is not based on UNIX or Linux?\nHow can I avoid using Facebook?\nHow do you get a Norton Security renewal code?\nDoes EMC plan a lightweight web client for Documentum?\nWhere is the strangest place you've ever masturba

In [ ]:
nltk.download('stopwords')
stop_words = stopwords.words('english')
print(stop_words)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
def TextPreproc (text: str, uniqs: bool=True):
  '''This function processes the incoming text: 
  tokenizes, removes punctuation and stop words, lemmatizes. 
  Returns a list of processed гтшйгуwords.'''
  text = text.lower()
  tokens_text = tokenizer.tokenize(text)
  words = [word for word in tokens_text if word.isalpha()]
  words = [w for w in words if not w in stop_words]
  func_lem = lambda w: lemmatizer.lemmatize(w)
  lem_words = list(map(func_lem, words))

  return (set(lem_words) if uniqs else lem_words)

In [ ]:
len(TextPreproc(first50))

213

**Вывод:** размер bag of words для 50 первых строк qoura после удаления стопслов и лематизации равен 213

# Дополнительное задание. Найти наиболее близкие вектора/строки среди первых 50 строк quora

In [ ]:
vector = TextPreproc(first50)

In [ ]:
matrix = []
for line in data[0:50]:
  prepoc_line = TextPreproc(line, False)
  temp_vector = []
  for i in vector:
    if i in prepoc_line:
      temp_vector.append(1)
    else:
      temp_vector.append(0)
  matrix.append(temp_vector)

In [ ]:
import numpy as np

In [ ]:
np.array(matrix).shape

(50, 213)

In [ ]:
cos_i_j = [0, 0, 0]
for i in range(len(matrix)-1):
  for j in range(i+1, len(matrix)):
    a = matrix[i]
    b = matrix[j]
    cos_sim = np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))
    if cos_sim > cos_i_j[0]:
      cos_i_j = [cos_sim, i, j]
print(cos_i_j[0])
print(cos_i_j[1], data[cos_i_j[1]])
print(cos_i_j[2], data[cos_i_j[2]])

0.5773502691896258
3 What are ZIP codes in the Bay Area?

11 What is the ZIP code of India?



**Вывод:** наиболее близкие строки под номером 3 и 11.